src/hotspot/share/runtime/handshake.cpp

```diff
@@ -569,6 +586,12 @@ bool HandshakeState::process_by_self(bool allow_suspend, bool check_async_except
   // set by this thread in case the operation is ThreadSuspendHandshake.
   OrderAccess::fence();
 
+#if INCLUDE_JFR
+  if (_handshakee->has_cpu_time_jfr_events()) {
+    Jfr::on_safepoint(_handshakee);
+  }
+#endif
```

Comment:
Adding this here plus all the other changes in the handshake code seem like a hack since this is not an actual handshake operation. I would just move this to `SafepointMechanism::process()` and add an additional condition in `update_poll_values()` and `should_process()`.